### PR TITLE
Merge v3 into v4

### LINKED ISSUE #40: Merge release 3.x into release 4.x
* What to do with the whole **extract** part?

### PR DESCRIPTION
**Pull request by 7ochem**
@erik-zicht can you check if the merge is ok? Don't check the code style-wise or anything. Just check if nothing will be changed/removed that shouldn't or something turned inconsistent or corrupt.

I'll also try to get this version `dev-merge-v3-into-v4` via Composer into a project and run it.

Resolves #40

### CODE REVIEW

File: src/Zicht/Bundle/SolrBundle/Manager/Doctrine/Subscriber.php [hunk 1]
```diff
@@ -20,7 +21,7 @@ class Subscriber implements EventSubscriber
      * @var ContainerInterface
      */
     protected $container;
```
Comment (7ochem) [inline src/Zicht/Bundle/SolrBundle/Manager/Doctrine/Subscriber.php]: This `$container` property has been made `protected` in v4, but in v3 it is `private`. I kept it `protected` (unchanged) during the merge as you can see. Now I get this error in a project where I'm upgrading it from v3.2.x to this merge branch:

> PHP Fatal error:  Access level to Zicht\Bundle\XxxyyyzzzSiteBundle\Solr\SiteSubscriber::$container must be protected (as in class Zicht\Bundle\SolrBundle\Manager\Doctrine\Subscriber) or weaker in src/Zicht/Bundle/XxxyyyzzzSiteBundle/Solr/SiteSubscriber.php on line 15

But you could expect to have to do some fixing when upgrading to a next major version...

File: src/Zicht/Bundle/SolrBundle/Manager/SolrManager.php [hunk 2]
```diff
@@ -41,19 +42,28 @@ class SolrManager
...
     public function addMapper($dataMapper)
     {
-        $this->mappers[]= $dataMapper;
+        $this->mappers[] = $dataMapper;
+    }
+
+    /**
+     * @return DataMapperInterface[]
+     */
+    public function getMappers()
```
Comment (7ochem) [inline src/Zicht/Bundle/SolrBundle/Manager/SolrManager.php]: This method is in here twice all of a sudden :see_no_evil:. Let me fix that :grin:

File: README.md [hunk 3]
```diff
@@ -7,5 +7,33 @@ Provides a layer to index and search content in a SOLR instance.
...
+
+### Mapping
+TODO: Write documentation about mapping documents to SOLR
+
+### Searching
+TODO: Write documentation about using SOLR for searching purposes.
+
+# Maintainers
+* Boudewijn Schoon <boudewijn@zicht.nl>
+* Philip Bergman <philip@zicht>
```
Comment (erik-zicht) [inline README.md]: Philip kan hier wel uit gehaald worden denk ik. Wie zijn de nieuwe maintainers?

Comment (7ochem) [reply README.md]: Ik zal het eens uitzoeken. Zal het wel in een volgende aanpassing doen. Dan kan dit iig door.

Deze hele bundle zit ook nog niet onder zicht standards, dus wordt ook nog niet gelint. Das ook nog een volgend puntje van aandacht :smile:

**Comment by erik-zicht** [comment]
I reviewed and tested the recent changes in v4 (reindex, batchupdate) and they seem to be fine and working properly in this branch. All of the extract-stuff I'm not familiar with but looks OK to me.

**Review by erik-zicht — APPROVED** [review]

**Pull request merged**